includes/plugins/wc-memberships/class-memberships.php

```diff
@@ -925,15 +925,16 @@ private static function user_has_content_access_from_rules( $user_id, array $rul
 
 	/**
 	 * Check if a user has an active subscription with the required products when checking membership status.
-	 * If they have an active subscription, reset inactive memberships to active link to the active subscription.
+	 * If they have an active subscription, but the membership is cancelled,
+	 * reset inactive memberships to active link to the active subscription.
 	 *
 	 * @param string  $post_status Post status.
 	 * @param WP_Post $post Post object.
 	 *
 	 * @return string
 	 */
 	public static function check_membership_status( $post_status, $post ) {
-		if ( 'wc_user_membership' !== $post->post_type || 'wcm-active' === $post->post_status || ! self::is_active() || ! function_exists( 'wc_memberships_get_user_membership' ) ) {
+		if ( 'wc_user_membership' !== $post->post_type || 'wcm-cancelled' !== $post->post_status || ! self::is_active() || ! function_exists( 'wc_memberships_get_user_membership' ) ) {
```

Comment:
There are other statuses that can be considered "inactive". I think to keep the original intention here this should read:

```suggestion
		if ( 'wc_user_membership' !== $post->post_type || ! in_array( $post->post_type, [ 'wcm-cancelled', 'wcm-expired', 'wcm-paused' ], true ) || ! self::is_active() || ! function_exists( 'wc_memberships_get_user_membership' ) ) {
```

Reply:
overall this is a very heavy routine to be executed in every `get_post_status` call... we should work on an alternative after we fix this...

Reply:
Do we know if any status other than `cancelled` (or other) is automatically set for memberships that got deactivated due to a subscription that failed to renew?

All status could be manually set (including cancelled), and would still get overwritten by this hook.

> overall this is a very heavy routine to be executed in every get_post_status call... we should work on an alternative after we fix this...

It bails early on `'wc_user_membership' !== $post->post_type`, I'm not sure it has a significant impact.